### ISSUE TITLE
New Chromium "Security" changes blocking TGS functionality cause of multiple open issues

### ISSUE THREAD

**Issue by CollinChaffin** [deanoemcke/thegreatsuspender#646]
This ram issue is absolutely 100% due to the recent changes in the Chromium code to address this big security whoopla that also if you notice is totally BREAKING this extension and much of the time (even when I build current source as of today) causing the icon I think folks are simply missing on the right side of their address bar for tabs that now just refuse to auto-suspend and if you hover that Chromium icon shows the "Redirect Blocked on this Page"......which if you click then opens a standard Chrome drop panel showing the "attempts" that TGS extension has made to "redirect" the tab (AKA to susepend it)........Screenshot:

In my debugging, THIS is what is causing Chrome to now chew up GIGS of RAM as it sits and fights what it now views as basically a security malware attempt to "redirect" when it really is just TGS background page doing what is has always (and who's soul purpose) has been.  Since all this hysteria between the CPU mfg's now throwing up to 20% of all our CPU power away, to all the browser and OS vendors doing these knee-jerk reaction changes without really thinking through the total impact like this, TGS is just one of many being the collateral damage.

In the case of Chromium, I haven't even tried to open any issue all I can think is that if extensions via our granting the varying levels of permissions ARE still supposed to be able to do what TGS is desiged for (being able to REDIRECT a tab), then I guess this should be considered a Chromium bug and I'm sure the TGS DEV team with it's huge user-base will have much more luck in getting it queued up and addressed by the Chromium DEV team.

EIther way, I apologize if this has already been identified but after trying to quickly go through all the open issues, I do see the RAM issue mentinoed (which i have referenced this new issue #) but did not see this detailed.

I am running multiple systems all reproducible with the following:

OS:  Windows 7 x64
Chrome:  65.0.3322.3 (Official Build) dev (64-bit)
TGS Entension: Was Chrome store latest (v6.30), but now running dev mode unpacked (v7.0.73)

It does seem like the FREQUENCY of Chromium blocking the TGS "redirects" is now changed with the latest version and what is odd is now DOES allow MANY to suspend, which is what makes the now-lower percentage that still are blocked by Chrome from suspending seem even more like a Chromium bug.

Screenshot of the issue (that then also causes issue #613 and Chrome RAM to go through the roof and continue to climb as it fights with the redirect attempts and when you have 200 tabs open where Chrome would already typically be using 6-8gb....and this occurs now it pretty much takes down the box LOL as referenced in #613   ):

![snag_1-19-2018_14-45-34](https://user-images.githubusercontent.com/9301483/35171128-44c9cdb0-fd28-11e7-9676-9ae6997540d1.png)

The OLD TGS version (Yes I reloaded and disabled it to address another open issue of various TGS versions not recognizing suspended tabs from other versions in saved sessions (#635 )) (Yes this was all tested with same results before re-loading old store version and then disabling - same results):

![snag_1-19-2018_14-46-47](https://user-images.githubusercontent.com/9301483/35171225-9e5d649a-fd28-11e7-8a8a-dcc8ea19d5fc.png)

The new Github version I built via source currently running and what rendered the above redirect error dropdown:

![snag_1-19-2018_14-46-16](https://user-images.githubusercontent.com/9301483/35171267-ccdf8140-fd28-11e7-83af-47e3ed39ca41.png)

FYI, not sure if related but there ARE in fact extension errors still being logged:

![snag_1-19-2018_14-46-31](https://user-images.githubusercontent.com/9301483/35171369-3bda84b4-fd29-11e7-99f9-5aea0877b220.png)

P.S.  Also, as a workaround for those with saved session either via other 3rd party extensions or TGS where TGS new versions will not recognize the suspension via old versions (root cause being the Chrome extension ID# changing) is this:  The "unsuspend ALL tabs" code must be doing a regex or other search (I haven't read the code) but despite the extension ID#'s changing between Github build etc........a single suspended page via diff version will NOT unsuspend via click so the trick is....go to any active tab (this is only PER WINDOW unfortunately so will have to be done for each WINDOW with mult tabs) and then click the TGS button and select the "unsuspend all tabs" and wholla it will then magically identify the old version's suspended open tabs and unsuspend them now allowing you to restore the original site's URL in the nav bar and take control of them with new TGS moving forward.

**Comment by deanoemcke** [comment]
@CollinChaffin can you please give me a bit more information on this?
I'm not seeing the blocked redirects you're speaking of.
And can you also post a link of the 'chrome changes' that you are referring to?
And finally, the version of chrome you currently use, and a working example of a website that fails to suspend properly.

Thanks :)

**Comment by bill-mcgonigle** [comment]
@deanoemcke , I think @CollinChaffin is referring to this:

![image](https://user-images.githubusercontent.com/880256/35889716-cf67b554-0b6a-11e8-8249-d959605e07a9.png)

Which I'm seeing on this URL:
chrome-extension://klbibkeccnjlkjkiokjodocebajanakg/suspended.html#ttl=systemd.target&uri=https://www.freedesktop.org/software/systemd/man/systemd.target.html

after upgrading to a github version.  @CollinChaffin, thanks for the workaround!

**Comment by deanoemcke** [comment]
@bill-mcgonigle I am unable to recreate this. What version of the extension are you using? And what version of chrome do you have? And finally, what operating system?

**Comment by deanoemcke** [comment]
@CollinChaffin Sorry, I just realised that you're on the chrome dev channel. But still, I cannot reproduce the blocking errors you're seeing.
As for the errors being logged to the console, that has been fixed in the latest version (v7.0.82).
Can somebody please give me more information on how i can reproduce the new blocking functionality?

**Comment by CollinChaffin** [comment]
Yes, it's this flag and a known issue past Chrome 62:

#enable-framebusting-needs-sameorigin-or-usergesture

I'm very surprised you're saying you cannot reproduce, since the new default after 62 is to kill the very redirection that GS relies on, LOL.  Leave this framebusting new partial attempt at built-in ad-blocking in Chrome, protects against GS redirects to the new suspension page, which is what I show here:

![snag_1-19-2018_14-45-34](https://user-images.githubusercontent.com/9301483/36272251-99690934-1246-11e8-8b21-7b5ad3bf9fbd.png)

To reproduce.  Ensure you're on latest Chrome DEV set the timer to few min and test it - even the latest Github build loaded unpacked 7.0.84, will fail to redirect SOME SOME SOME.  That is the issue - this is Chrome's new AI so sometime's the tabs suspend.  It is funny that one of the ones that seems to almost NEVER suspend, is www.google.com.  Open the main google search, perform a search, then switch to a different tab and wait the 15 min and you will see that tab will NOT suspend.  Go back to visit it and look on the right hand side of the omnibar and you will see this screenshot.  Disabling this "feature" via the flag I just posted here fixes the issue (for now until they force this feature on later).

**Comment by CollinChaffin** [comment]
More info for you on this new "feature" and flag:

https://www.androidpolice.com/2017/06/01/google-confirms-will-block-ads-non-compliant-sites-chrome-starting-early-2018/
https://www.androidpolice.com/2017/11/08/heres-protect-rogue-redirect-ads-right-now-chrome/
http://www.wired.co.uk/article/google-chrome-ad-blocker-filter-how-works-coalition-ads

Either way looks like you'll have to rewrite the way you redirect in Chrome.....eventually.  And instruct your users to disable this ad-blocking flag if they expect GS to work for the time being. :)

**Comment by CollinChaffin** [comment]
Just an update as of 03-06-2018 on Version 66.0.3355.0 (Official Build) dev (64-bit):

These issues still exist.  Clearly Google in implementing their own new "let's fight all redirects" code more than just the above flag which seemed to help for a bit and now wonder if it was until DEV channel updated again because now, it almost seems worse like almost 50% of my tabs do not suspend and result in the above "redirect blocked", whereas the other 50% for some reason Chrome decides won the lottery and allows a nice quiet redirect the way it always used to work.

EDIT:  I just realized I am a few commits behind so I am pulling down the repo now to install the very latest code here to see if some of the very recent refactoring and fixes have had any impact.

**Comment by deanoemcke** [comment]
@CollinChaffin I just can't seem to replicate this issue.
Here are the exact steps I have tried:
- Chrome Dev:  v66.0.3359.26
- Ensure flag: #enable-framebusting-needs-sameorigin-or-usergesture is enabled
- Extension version: v7.0.85 (I have also tried v6.30)
- Open about 30 random tabs (many of them google search tabs)
- Wait for them to autosuspend (i have it set to 20 seconds)
- Ensure that they all suspend automatically

What do you think I'm doing differently to you? 
I cannot get that popup blocked icon to appear that you have posted an image of.

**Comment by CollinChaffin** [comment]
How is that possible?  There are at least 3-4 other open TGS issues right now (#661 referenced above) all having this exact issue and I literally have a DOZEN machines all on v66 Chrome and 100% of them exhibit this.  In fact, I have just opened a Chromium issue as well at:
````
https://bugs.chromium.org/p/chromium/issues/detail?id=822791
````
**Anyone else also experiencing this that feels it needs DEV attention should visit that issue and STAR it so that it gains some DEV attention.

As this can easily be reproduced on a large scale.  In fact, I thought I already shot a video screen-cap for you, but if I have time later and this weekend I will be recording some actual video footage of this behavior, both for you here as well as the Chromium DEVs because this new anti-redirect code is just simply not ready for prime-time.  As I indicated in my Chrome issue, I literally downloaded 20+ free Solarwinds files, all from the exact same open tab in the same session etc. and only after a DOZEN normal downloads, did Chrome v66 suddenly out of nowhere (URLs to zips also did not change) began totally BLOCKING the download and placing the very same redirect block on the omni-bar.  Upon dropping down the warning on the omni-bar, the zip file url is shown and clicking it then proceeds to the download the zip just as it did the prior 12 items without Chrome feeling the need to "protect" me, LOL.

**Comment by CollinChaffin** [comment]
EDIT:  I am now re-testing after you post (and after this fiasco of the push to everyone of Chrome DEV v66.0.3359.23 that rendered it totally unable to launch), and finally getting my systems to jump past to v66.0.3359.26, I had not admittedly re-checked every flag (they SHOULD stick), but not only is #enable-framebusting-needs-sameorigin-or-usergesture no longer enabled on these systems, but I see other flags like the experimental new V8 PARSER optimization have flat out now been removed and instead of falling back to the next best thing (COMPILER optimized), totally fell back to DISABLED as well. :(

I'm curious in my open Chromium issue if they will confirm whether that flag really is supposed to be totally disabling this new redirection blocking or not.  I see several blog posts from Chromium DEVs from a while back that seem to indicate that this new feature is still in development and might not block all types of redirects yet so maybe the code really has not been fully implemented yet after all, it is a DEV build. :)

I just really don't like it when it's this bad and on deck to hit the stable channel so soon so hopefully they get it worked out and TGS won't have to code around it, other than to perhaps indicate the user must set a flag etc.

**Comment by deanoemcke** [comment]
Don't worry. You have my full attention, you don't need to garner extra community support. I am treating this as a major potential issue. Having users manually setting chrome flags is not an option. If this is going to affect all users in an upcoming release, then a fixed version of the extension needs to be released BEFORE this happens.

However, I am still unable to reproduce. I'm really sorry.
My current setup is:
- Chrome Dev Version 66.0.3359.26
- TGS v6.30
- Framebusting flag still enabled

I have tried going to your https://www.solarwinds.com/downloads page and opening every trial box in a new tab, and from there clicking on all the download links after filling out the form.
When downloading the files, I did initially get a popup blocked notification which i then allowed (by turning of the block popups option in chrome). After this the files were able to download successfully.
At no point did I get a 'redirect blocked' notification. 
For good measure I then tried suspending all the open tabs and they all suspended without issue.

Perhaps this has something to do with the fact that I do not use chrome dev very often? But even then, my production chrome also has the framebusting flag set, and I'm not seeing issues there either.
